### PR TITLE
Fix performance issue with shift operations

### LINKED ISSUE #13: Performance of bit shift
Hello, great package!

Is this kind of performance regression expected?

```julia
julia> @btime $(Ref(Int1024(2131)))[] << $(Ref(1))[]
  295.443 ns (0 allocations: 0 bytes)
4262

julia> @btime $(Ref(Int64(2131)))[] << $(Ref(1))[]
  2.256 ns (0 allocations: 0 bytes)
4262
```

### PR DESCRIPTION
**Pull request by jessymilare**
Fixes #13

### CODE REVIEW

**Comment by rfourquet** [comment]
Thanks! I will try to review in the next few days.

File: test/runtests.jl [hunk 1]
```diff
@@ -194,6 +194,24 @@ end
 end
 
 
+@testset "shift operations" begin
+    vals = [rand(Int8, 8)..., rand(Int64, 8)..., rand(Int1024, 8)...]
+    shifts = [rand(-8:8, 8)..., rand(-64:64, 8)..., rand(-1024:1024, 8)...]
```
Comment (rfourquet) [inline test/runtests.jl]: While you are at it, would you also test for non-Int shifts ? e.g.
 ```suggestion
    shifts = [rand(-8:8, 8)..., rand(-64:64, 8)..., rand(Int256(-1024):Int256(1024), 8)...]
```

File: test/runtests.jl [hunk 2]
```diff
@@ -194,6 +194,24 @@ end
 end
 
 
+@testset "shift operations" begin
```
Comment (rfourquet) [inline test/runtests.jl:197]: Wow, I had no idea shifts were almost not tested!

File: src/BitIntegers.jl [hunk 3]
```diff
@@ -311,14 +311,40 @@ promote_rule(::Type{Float16}, ::Type{<:XBI}) = Float16
...
->>>(x::UBI, y::Int) = ifelse(0 <= y, x >>> unsigned(y), x << unsigned(-y))
+# LLVM bit shifting (lshr_int, ashr_int and shl_int) has performance issues for large integers
+# The following functions allows much faster code (10x speedup for 1024 bit integers)
+
+# Note: this should be a power of 2
+const SHIFT_SPLIT_NBITS = 64
+
+@inline @generated function shift_call(sh_fun::Function, x::I, y::UBU) where {I<:XBI}
+    nbits = 8 * sizeof(I)
+    # Performance issue does not afect integers with less than 256 bits
```
Comment (rfourquet) [inline src/BitIntegers.jl]:
```suggestion
    # performance issue does not affect integers with no more than 128 bits
```
(we can have e.g. `Int200` in between)

File: src/BitIntegers.jl [hunk 4]
```diff
@@ -311,14 +311,40 @@ promote_rule(::Type{Float16}, ::Type{<:XBI}) = Float16
...
->>>(x::UBI, y::UBU) = lshr_int(x, y)
-<<( x::UBI, y::UBU) = shl_int(x, y)
-
->>( x::UBI, y::Int) = ifelse(0 <= y, x >> unsigned(y),  x << unsigned(-y))
-<<( x::UBI, y::Int) = ifelse(0 <= y, x << unsigned(y),  x >> unsigned(-y))
->>>(x::UBI, y::Int) = ifelse(0 <= y, x >>> unsigned(y), x << unsigned(-y))
+# LLVM bit shifting (lshr_int, ashr_int and shl_int) has performance issues for large integers
+# The following functions allows much faster code (10x speedup for 1024 bit integers)
+
+# Note: this should be a power of 2
```
Comment (rfourquet) [inline src/BitIntegers.jl:317]: So must it be so because shifting by such constant values is particularly efficient? or will `shift_call` also give wrong answers?

File: src/BitIntegers.jl [hunk 5]
```diff
@@ -311,14 +311,40 @@ promote_rule(::Type{Float16}, ::Type{<:XBI}) = Float16
...
+@inline @generated function shift_call(sh_fun::Function, x::I, y::UBU) where {I<:XBI}
+    nbits = 8 * sizeof(I)
+    # Performance issue does not afect integers with less than 256 bits
+    nbits <= 128 && return :(sh_fun(x, y))
+    split = SHIFT_SPLIT_NBITS
+    mask = split - 1
+    quote
+        hi = y >> $(trailing_zeros(split))
+        $([:(hi == $val && return sh_fun(sh_fun(x, $(val * split)), y & $mask))
+            for val in 0:div(nbits, split)]...
```
Comment (rfourquet) [inline src/BitIntegers.jl:329]: This looks correct, but I think I would feel a bit more comfortable merging this if you added a couple more tests testing edge cases, e.g. when `val` doesn't fall in `0:div(nbits, split)]`, or when it's equal to `div(nbits, split)]`, or with e.g. `Int200`. I know this was not tested before, this is now a more complicated algorithm than what was in `Base` (i.e. using directly intrinsics).

**Review by rfourquet — APPROVED** [review]
Thanks, that's great! I don't have time right now to test performance myself, so I will just merge, we can always incrementally improve if we find cases where this is not an improvement. It would be awesome if you can add a couple more tests as I mention in a comment, otherwise I will add some eventually before tagging a release.

File: src/BitIntegers.jl (hunk 4 above)
Comment (jessymilare) [reply src/BitIntegers.jl:317]: LLVM generates a correct code for shifting. It is slow in general because it emits code for several shift lengths and, instead of testing and jumping into the correct branch, it runs all branches and selects the correct result later. When the second argument is constant, the emitted code is specialized for that shift length (it eliminates useless branches). The same happens when the second argument is known by LLVM to be smaller than 64.

The only reason that `SHIFT_SPLIT_NBITS` should be a power of 2 is that the code splits `y` into lower (`y & mask`) and higher (`hi`) bits. I haven't tested in 32-bit machines, it could be better to set `SHIFT_SPLIT_NBITS = 32` in that case (I'm unsure, though).

I haven't found any instance in which the code generated is incorrect.

File: src/BitIntegers.jl (hunk 3 above)
Comment (jessymilare) [reply src/BitIntegers.jl]: I actually didn't test the performance for integers between `Int128` and `Int256`, so I'm not sure if `Int200` would have performance issue.

File: src/BitIntegers.jl (hunk 5 above)
Comment (jessymilare) [reply src/BitIntegers.jl:329]: I agree. I added many more tests in the last commit, it should cover all cases you mentioned.

File: src/BitIntegers.jl (hunk 3 above)
Comment (rfourquet) [reply src/BitIntegers.jl]: Ok, so I think my suggestion makes sense then? As the fast-path you write is for `nbits <= 128`.
